src/mongo_types.ts

```diff
@@ -42,34 +42,145 @@ export type EnhancedOmit<TRecordOrUnion, KeyUnion> = string extends keyof TRecor
 export type WithoutId<TSchema> = Omit<TSchema, '_id'>;
 
 /** A MongoDB filter can be some portion of the schema or a set of operators @public */
-export type Filter<TSchema> = Partial<TSchema> & Document;
+export type Filter<TSchema> = {
+  [P in keyof TSchema]?: Condition<TSchema[P]>;
+} &
+  RootFilterOperators<TSchema>;
 
-/** A MongoDB UpdateQuery is set of operators @public */
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
-export type UpdateQuery<TSchema> = Document; // TODO(NODE-3274)
+/** @public */
+export type Condition<T> = AlternativeType<T> | FilterOperators<AlternativeType<T>>;
+
+/**
+ * It is possible to search using alternative types in mongodb e.g.
+ * string types can be searched using a regex in mongo
+ * array types can be searched using their element type
+ * @public
+ */
+export type AlternativeType<T> = T extends ReadonlyArray<infer U>
+  ? T | RegExpOrString<U>
+  : RegExpOrString<T>;
+
+/** @public */
+export type RegExpOrString<T> = T extends string ? BSONRegExp | RegExp | T : T;
+
+/** @public */
+export interface RootFilterOperators<TSchema> extends Document {
+  $and?: Filter<TSchema>[];
+  $nor?: Filter<TSchema>[];
+  $or?: Filter<TSchema>[];
+  $text?: {
+    $search: string;
+    $language?: string;
+    $caseSensitive?: boolean;
+    $diacriticSensitive?: boolean;
+  };
+  $where?: string | ((this: TSchema) => boolean);
+  $comment?: string | Document;
+}
+
+/** @public */
+export interface FilterOperators<TValue> extends Document {
+  // Comparison
+  $eq?: TValue;
+  $gt?: TValue;
+  $gte?: TValue;
+  $in?: TValue[];
+  $lt?: TValue;
+  $lte?: TValue;
+  $ne?: TValue;
+  $nin?: TValue[];
+  // Logical
+  $not?: TValue extends string ? FilterOperators<TValue> | RegExp : FilterOperators<TValue>;
+  // Element
+  /**
+   * When `true`, `$exists` matches the documents that contain the field,
+   * including documents where the field value is null.
+   */
+  $exists?: boolean;
+  $type?: BSONType | BSONTypeAlias;
+  // Evaluation
+  $expr?: any;
+  $jsonSchema?: any;
+  $mod?: TValue extends number ? [number, number] : never;
+  $regex?: TValue extends string ? RegExp | BSONRegExp | string : never;
+  $options?: TValue extends string ? string : never;
+  // Geospatial
+  $geoIntersects?: { $geometry: Document };
+  $geoWithin?: Document;
+  $near?: Document;
+  $nearSphere?: Document;
+  $maxDistance?: number;
+  // Array
+  $all?: TValue extends ReadonlyArray<any> ? any[] : never;
+  $elemMatch?: TValue extends ReadonlyArray<any> ? Document : never;
+  $size?: TValue extends ReadonlyArray<any> ? number : never;
+  // Bitwise
+  $bitsAllClear?: BitwiseFilter;
+  $bitsAllSet?: BitwiseFilter;
+  $bitsAnyClear?: BitwiseFilter;
+  $bitsAnySet?: BitwiseFilter;
```

Comment:
Just missing `$rand` that is an empty document if we want to be fully complete in this list.